kernel/poll.c

```diff
@@ -188,36 +188,77 @@ static inline void set_event_ready(struct k_poll_event *event, u32_t state)
 	event->state |= state;
 }
 
-int z_impl_k_poll(struct k_poll_event *events, int num_events, s32_t timeout)
+static inline int register_events(struct k_poll_event *events,
+				  int num_events,
+				  struct _poller *poller,
+				  bool just_check)
 {
-	__ASSERT(!z_arch_is_in_isr(), "");
-	__ASSERT(events != NULL, "NULL events\n");
-	__ASSERT(num_events > 0, "zero events\n");
-
-	int last_registered = -1, rc;
-	k_spinlock_key_t key;
+	int events_registered = 0;
 
-	struct _poller poller = { .thread = _current, .is_polling = true, };
-
-	/* find events whose condition is already fulfilled */
 	for (int ii = 0; ii < num_events; ii++) {
+		k_spinlock_key_t key;
 		u32_t state;
 
 		key = k_spin_lock(&lock);
 		if (is_condition_met(&events[ii], &state)) {
 			set_event_ready(&events[ii], state);
-			poller.is_polling = false;
-		} else if (timeout != K_NO_WAIT && poller.is_polling) {
-			rc = register_event(&events[ii], &poller);
+			poller->is_polling = false;
+		} else if (!just_check && poller->is_polling) {
+			int rc = register_event(&events[ii], poller);
 			if (rc == 0) {
-				++last_registered;
+				events_registered += 1;
 			} else {
 				__ASSERT(false, "unexpected return code\n");
 			}
 		}
 		k_spin_unlock(&lock, key);
 	}
 
+	return events_registered;
+}
+
+static int k_poll_poller_cb(struct k_poll_event *event, u32_t state)
+{
+	struct k_thread *thread = event->poller->thread;
+
+	__ASSERT(thread != NULL, "poller should have a thread\n");
+
+	if (!z_is_thread_pending(thread)) {
+		return 0;
+	}
+
+	if (z_is_thread_timeout_expired(thread)) {
+		return -EAGAIN;
+	}
+
+	z_unpend_thread(thread);
+	z_arch_thread_return_value_set(thread,
+		state == K_POLL_STATE_CANCELLED ? -EINTR : 0);
+
+	if (!z_is_thread_ready(thread)) {
+		return 0;
+	}
+
+	z_ready_thread(thread);
+
+	return 0;
+}
+
+int z_impl_k_poll(struct k_poll_event *events, int num_events, s32_t timeout)
+{
+	int events_registered;
+	k_spinlock_key_t key;
+	struct _poller poller = { .is_polling = true,
+				  .thread     = _current,
+				  .cb         = k_poll_poller_cb };
+
+	__ASSERT(!z_is_in_isr(), "");
```

Comment:
Sorry, didn't run it first: this needs to be `z_arch_is_in_isr()` (upstream rename).

Reply:
Fixed.